torch/csrc/fx/node.cpp

```diff
@@ -0,0 +1,269 @@
+    return false;
+  }
+  Py_INCREF(&NodeBaseType);
+  if (PyModule_AddObject(module, "_NodeBase", (PyObject*)&NodeBaseType) != 0) {
+    return false;
+  }
+  return true;
+}
+
+////////////////////////////////
+// NodeIter
+////////////////////////////////
+
+struct NodeIter {
+  PyObject_HEAD bool _reversed;
+  NodeBase* _root;
+  NodeBase* _cur;
+};
+
+static PyObject* NodeIter_new(
+    PyTypeObject* type,
+    PyObject* args,
+    PyObject* kwds) {
+  PyObject* self = type->tp_alloc(type, 0);
+  if (!self)
+    return nullptr;
+  return self;
+}
+
+static int NodeIter_init_fn(NodeIter* self, PyObject* args, PyObject* kwargs) {
+  NodeBase* root = nullptr;
+  bool reversed = false;
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays,modernize-avoid-c-arrays)
+  constexpr const char* keywords[] = {"root", "reversed", nullptr};
+  if (!PyArg_ParseTupleAndKeywords(
+          args,
+          kwargs,
+          "Ob|",
+          // NOLINTNEXTLINE(cppcoreguidelines-pro-type-const-cast)
+          const_cast<char**>(keywords),
+          &root,
+          &reversed)) {
+    return -1;
+  }
+  self->_reversed = reversed;
+  Py_INCREF(root);
+  self->_root = root;
+  Py_INCREF(root);
+  self->_cur = root;
+  return 0;
+}
+
+PyObject* NodeIter_iter(PyObject* self) {
+  Py_INCREF(self);
+  return self;
+}
+
+template <bool reversed>
+PyObject* NodeIter_iternext_helper(NodeIter* self) {
+  // It should be possible to relax the ref counting here
+  // but in practice, we do not have that many _erased Nodes,
+  // so probably not worth it.
+  if constexpr (reversed) {
+    Py_INCREF(self->_cur->_prev);
+    Py_XDECREF(self->_cur);
+    self->_cur = self->_cur->_prev;
+  } else {
+    Py_INCREF(self->_cur->_next);
+    Py_XDECREF(self->_cur);
+    self->_cur = self->_cur->_next;
+  }
+  while (self->_cur != self->_root) {
+    if (!self->_cur->_erased) {
+      Py_INCREF(self->_cur);
+      return (PyObject*)self->_cur;
+    }
+    if constexpr (reversed) {
+      Py_INCREF(self->_cur->_prev);
+      Py_XDECREF(self->_cur);
+      self->_cur = self->_cur->_prev;
+    } else {
+      Py_INCREF(self->_cur->_next);
+      Py_XDECREF(self->_cur);
+      self->_cur = self->_cur->_next;
+    }
+  }
+  PyErr_SetNone(PyExc_StopIteration);
+  return nullptr;
+}
+
+PyObject* NodeIter_iternext(PyObject* _self) {
+  NodeIter* self = (NodeIter*)_self;
+  if (self->_reversed) {
+    return NodeIter_iternext_helper<true>(self);
+  } else {
+    return NodeIter_iternext_helper<false>(self);
+  }
+}
+
+static int NodeIter_traverse(NodeIter* self, visitproc visit, void* arg) {
+  Py_VISIT(self->_root);
+  Py_VISIT(self->_cur);
+  return 0;
+}
+
+static int NodeIter_clear(NodeIter* self) {
+  Py_CLEAR(self->_root);
+  Py_CLEAR(self->_cur);
+  return 0;
+}
+
+static void NodeIter_dealloc(PyObject* self) {
+  PyObject_GC_UnTrack(self);
+  (void)NodeIter_clear((NodeIter*)self);
+  Py_TYPE(self)->tp_free(self);
+}
+
+static PyTypeObject NodeIterType = {
```

Comment:
We're not doing the binding in pybind11 because speed is paramount

Reply:
But my guess is that the main speedup here is to do the iterations via direct inline access instead of a full python opcode every time. Is the actual cost of calling from python->c++ that big here?
I am not suggesting we should rewrite this PR that already exists! I was just curious if we tried it and if so what were the numbers.

Reply:
It just seems weird to me to say, "Hey, for this microoptimization, have you tried benchmarking it with X, where X is something we *know* is really slow." Like hey, if your goal is to speed things up, don't do things you know are slow!!

Reply:
Yep, doing this via PyBind would completely defeat this PR

Reply:
But if you get 95% of the speedup with 1/3 of the code, it takes you half the time to land and it is less likely to contain bugs that sounds like a valid tradeoff to consider? Am I missing something here?

Reply:
I'm very doubtful that you'll get 95% of the speedup. The overhead of calling into pybind will likely even turn this into a regression. @jansel ?

Since I already implemented this, I don't really want to implement it again in pybind just so that we can prove that it is much worse in performance :/

Reply:
> Since I already implemented this, I don't really want to implement it again in pybind just so that we can prove that it is much worse in performance :/

Of course, as I said above, this was really not what I was asking for here.
I was asking in case you had that data so I could improve my mental model of how, when and by how much is pybind slower than manual bindings.